### ISSUE TITLE
Preparing for PHP 7.0

### ISSUE THREAD

**Issue by silverkorn** [phalcon/zephir#756]
Hey there!
Do you plan to add support to the PHP NG (aka _Zend engine 3_) in a near future?

The PHP team is planning to release it around __Mid-October 2015__ and probably freeze major changes during Mid-June according to this link: 
https://wiki.php.net/rfc/php7timeline.

Here's the documentation for upgrading from Zend engine 2 to PHP NG (subject to change a few until development freeze): https://wiki.php.net/phpng-upgrading along with https://wiki.php.net/phpng-int.

I guess a new branch could be make to prepare with those changes to hopefully synchronize (or at least being nearly ready) to their release date.

Thanks!

**Comment by ovr** [comment]
Yeap, We planned to support PHP 7, and next versions but not now.

**Comment by silverkorn** [comment]
Great! Thanks for the confirmation.

**Comment by steffengy** [comment]
@ovr 
I already have a version, which provides basic functionality against php-src (current PHP 7 state - master branch).
The differences are huge and I'm not quite sure if providing backwards compatibility is even possible, without adding very much additional boilerplate-versionchecking code,
but nevertheless an update until release should be possible.
(If they do not make some more fundamental changes)

**Comment by loint** [comment]
:+1:  great to hear that. We 're using zephir as our main language :)

**Comment by steffengy** [comment]
I will probably put my experiment on github, when all tests pass.
(Except if that's not desired)

Current state (with not all tests enabled yet):
```
OK, but incomplete, skipped, or risky tests!
Tests: 89, Assertions: 509, Risky: 8.
```

Hopefully I'll manage to get this done, since it will probably be a helpful 
reference when updating zephir to PHP 7.

**Comment by steffengy** [comment]
https://github.com/steffengy/zephir/tree/zephir-ngexp

It is far from perfect, but as an initial attempt pretty good.
It runs **(pretty much) all tests on win32** 
(except the interface-constant-inheritance/preg_match as described
in the readme)
and I'm pretty satisfied for an initial attempt (a week of work).

I will probably not continue this experiment for now.
The **linux tests on travis somehow unfortunately fail**.
This could fail because **zephir compiler is not yet
compatible with PHP 7 yet** (atleast it crashes on windows).

Feel free to use it as reference when preparing for PHP 7, 
if it shouldn't have changed significantly until then.

**Comment by ovr** [comment]
Good job @steffengy :+1:

**Comment by steffengy** [comment]
@ovr
Travis build now also passes as sucessful!
(Fixed some (annoying) heap corruption)

Furthermore I encountered the following problems, while conducting this experiment:
- **Limited tests**

The zephir tests only check for very basic functionality, ignoring complex programs.
- **Code duplication**

The compiler contains much code duplication, requiring more adjustments.
Also it does affect readability.
Also there's room for improvement in the "kernel sector", since some stuff is just copy & pasted
and existing multiple times.

Also on a long term, zephir should generate more efficient code, like reusing temp variables or performing more optimizations on the generated code.

**Comment by loint** [comment]
Yeah, let me try it :+1:

**Comment by steffengy** [comment]
@loint there probably is not much to try, since large-scale projects likely will not work correctly.
I will probably test with something pretty big,
when I have more spare time.
If you want to help, implementing more complex unit tests is probably more effective :)
Nevertheless testing stuff is never bad and could help too - please let me know what your testings show :)

**Comment by loint** [comment]
yes, it's interesting . I will do it later . :)

**Comment by steffengy** [comment]
@andresgutierrez 
Is there already an official plan/timeline for this feature?

Also do you already have an idea how to ensure backwards-compatibility/dual-implementation for
all the functions using ``zval **`` (which should be ``zval *`` if done the same way as in PHP7) ?
(since that is pretty much not used anymore in PHP7 - 
or will zephir regardless of the rest of the PHP-Code still implement it that way?)

Also ``ALLOC_INIT_ZVAL`` does not exist anymore and zval's are allocated
mostly on stack, which also may require different code generation.
(Is there an easy way to do this? How will that work with function calls, which have incosistent checks for "stack-only" variables (isLocalOnly - which then most is?)
TLDR: Code-generation results in chaos and would need more checks/abstraction

Kernel-wise it also seems like you'll have very much ifdef's in the code.
Maybe having multiple versions of files for < 7 and >= 7 makes sense for some files?
(with guards which check the PHP_VERSION)

Feel free to take a look at the experiment I conducted earlier (mentioned somewhere above),
where I tried to migrate everything. (No BC - the code is far from working properly at some places though)

Nevertheless it will be much work, do you see anything what would prevent
preparations/start of the work (which needs this initial planning, which unfortunately
won't be trivial)

**Comment by andresgutierrez** [comment]
Closing this, leaving #952

**Comment by masums** [comment]
I have tried to compile using PHP 7 but shows following error.
ZVAL_STRING takes parameter 2 but 3 passed.

php info

PHP 7.0.0RC5 (cli) (built: Oct 25 2015 17:36:08) ( NTS )
Copyright (c) 1997-2015 The PHP Group
Zend Engine v3.0.0-dev, Copyright (c) 1998-2015 Zend Technologies

**Comment by andresgutierrez** [comment]
@masums Are you compiling a custom Zephir project?

**Comment by masums** [comment]
I have compiled at RedHat Linux terminal where PHP 7 RC5 was installed.

On Sat, Oct 31, 2015 at 12:58 AM, Andy Gutierrez notifications@github.com
wrote:

> @masums https://github.com/masums Are you compiling a custom Zephir
> project?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/phalcon/zephir/issues/756#issuecomment-152617085.

## 

Md. Rashidul Hasan Masum http://masums.com
Software Architect
OnnoRokom Software Ltd. http://onnorokomsoftware.com
Mobile: 88018 11 41 95 57

**Comment by r3ndyipb** [comment]
i have same issue with @masums 

configure: error: in `/root/cphalcon/ext':
configure: error: C compiler cannot create executables

i use centos 7 and php 7 RC5

**Comment by steffengy** [comment]
That sounds like a toolchain problem, which is not related to any PHP version or zephir version,
it just shows the lack of an appropriate tooling on your system.

**Comment by r3ndyipb** [comment]
sory i forget install make and gcc :D 

but i found this error when i do this

git clone http://github.com/phalcon/cphalcon
git checkout 2.0.x
cd ext
sudo ./install

error :

libtool: compile:  gcc -I. -I/root/cphalcon/ext -DPHP_ATOM_INC -I/root/cphalcon/ext/include -I/root/cphalcon/ext/main -I/root/cphalcon/ext -I/usr/include/php -I/usr/include/php/main -I/usr/include/php/TSRM -I/usr/include/php/Zend -I/usr/include/php/ext -I/usr/include/php/ext/date/lib -DHAVE_CONFIG_H -O2 -Wall -fvisibility=hidden -flto -DZEPHIR_RELEASE=1 -c /root/cphalcon/ext/phalcon.c  -fPIC -DPIC -o .libs/phalcon.o
In file included from /root/cphalcon/ext/phalcon.h:15:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator.zep.h:12:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator(zend_class_entry _class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:125:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/assets/collection.zep.h:41:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Assets_Collection(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:223:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/http/response/headers.zep.h:14:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Http_Response_Headers(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:288:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/email.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Email(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:289:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/exclusionin.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Exclusionin(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:290:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/inclusionin.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Inclusionin(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:291:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/ip.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Ip(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:292:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/numericality.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Numericality(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:293:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/presenceof.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_PresenceOf(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:294:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/regex.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Regex(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:295:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/stringlength.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_StringLength(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:296:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/uniqueness.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Uniqueness(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:297:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/url.zep.h:7:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Url(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.h:336:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/translate/adapter/csv.zep.h:10:1: error: unknown type name 'zend_object_value'
 static zend_object_value zephir_init_properties_Phalcon_Translate_Adapter_Csv(zend_class_entry *class_type TSRMLS_DC);
 ^
In file included from /root/cphalcon/ext/phalcon.c:24:0:
/root/cphalcon/ext/kernel/main.h: In function 'zephir_maybe_separate_zval':
/root/cphalcon/ext/kernel/main.h:91:2: warning: implicit declaration of function 'Z_REFCOUNT_PP' [-Wimplicit-function-declaration]
  if (Z_REFCOUNT_PP(z) > 1 && !Z_ISREF_PP(z)) {
  ^
/root/cphalcon/ext/kernel/main.h:91:2: warning: implicit declaration of function 'Z_ISREF_PP' [-Wimplicit-function-declaration]
/root/cphalcon/ext/kernel/main.h:94:3: warning: implicit declaration of function 'ALLOC_ZVAL' [-Wimplicit-function-declaration]
   ALLOC_ZVAL(new_zv);
   ^
/root/cphalcon/ext/kernel/main.h:95:3: warning: implicit declaration of function 'Z_UNSET_ISREF_P' [-Wimplicit-function-declaration]
   INIT_PZVAL_COPY(new_zv, *z);
   ^
In file included from /root/cphalcon/ext/phalcon.c:25:0:
/root/cphalcon/ext/kernel/fcall.h: In function 'zephir_return_call_function':
/root/cphalcon/ext/kernel/fcall.h:586:3: warning: passing argument 1 of '_zval_ptr_dtor' from incompatible pointer type [enabled by default]
   zval_ptr_dtor(return_value_ptr);
   ^
In file included from /usr/include/php/Zend/zend.h:39:0,
                 from /usr/include/php/main/php.h:35,
                 from /root/cphalcon/ext/phalcon.c:8:
/usr/include/php/Zend/zend_variables.h:112:15: note: expected 'struct zval *' but argument is of type 'struct zval *_'
 ZEND_API void _zval_ptr_dtor(zval _zval_ptr ZEND_FILE_LINE_DC);
               ^
In file included from /root/cphalcon/ext/phalcon.c:25:0:
/root/cphalcon/ext/kernel/fcall.h:594:4: warning: implicit declaration of function 'ALLOC_INIT_ZVAL' [-Wimplicit-function-declaration]
    ALLOC_INIT_ZVAL(_return_value_ptr);
    ^
/root/cphalcon/ext/kernel/fcall.h:601:3: warning: implicit declaration of function 'COPY_PZVAL_TO_ZVAL' [-Wimplicit-function-declaration]
   COPY_PZVAL_TO_ZVAL(_return_value, rv);
   ^
/root/cphalcon/ext/kernel/fcall.h: In function 'zephir_return_call_zval_function':
/root/cphalcon/ext/kernel/fcall.h:623:3: warning: passing argument 1 of '_zval_ptr_dtor' from incompatible pointer type [enabled by default]
   zval_ptr_dtor(return_value_ptr);
   ^
In file included from /usr/include/php/Zend/zend.h:39:0,
                 from /usr/include/php/main/php.h:35,
                 from /root/cphalcon/ext/phalcon.c:8:
/usr/include/php/Zend/zend_variables.h:112:15: note: expected 'struct zval *' but argument is of type 'struct zval *_'
 ZEND_API void _zval_ptr_dtor(zval _zval_ptr ZEND_FILE_LINE_DC);
               ^
In file included from /root/cphalcon/ext/phalcon.c:25:0:
/root/cphalcon/ext/kernel/fcall.h: In function 'zephir_return_call_class_method':
/root/cphalcon/ext/kernel/fcall.h:664:3: warning: passing argument 1 of '_zval_ptr_dtor' from incompatible pointer type [enabled by default]
   zval_ptr_dtor(return_value_ptr);
   ^
In file included from /usr/include/php/Zend/zend.h:39:0,
                 from /usr/include/php/main/php.h:35,
                 from /root/cphalcon/ext/phalcon.c:8:
/usr/include/php/Zend/zend_variables.h:112:15: note: expected 'struct zval *' but argument is of type 'struct zval *_'
 ZEND_API void _zval_ptr_dtor(zval _zval_ptr ZEND_FILE_LINE_DC);
               ^
In file included from /root/cphalcon/ext/phalcon.h:15:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon.c: At top level:
/root/cphalcon/ext/phalcon/mvc/model/validator.zep.h:12:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:125:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/assets/collection.zep.h:41:26: warning: 'zephir_init_properties_Phalcon_Assets_Collection' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Assets_Collection(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:223:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/http/response/headers.zep.h:14:26: warning: 'zephir_init_properties_Phalcon_Http_Response_Headers' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Http_Response_Headers(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:288:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/email.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Email' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Email(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:289:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/exclusionin.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Exclusionin' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Exclusionin(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:290:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/inclusionin.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Inclusionin' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Inclusionin(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:291:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/ip.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Ip' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Ip(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:292:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/numericality.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Numericality' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Numericality(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:293:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/presenceof.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_PresenceOf' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_PresenceOf(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:294:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/regex.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Regex' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Regex(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:295:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/stringlength.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_StringLength' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_StringLength(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:296:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/uniqueness.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Uniqueness' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Uniqueness(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:297:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/mvc/model/validator/url.zep.h:7:26: warning: 'zephir_init_properties_Phalcon_Mvc_Model_Validator_Url' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Mvc_Model_Validator_Url(zend_class_entry *class_type TSRMLS_DC);
                          ^
In file included from /root/cphalcon/ext/phalcon.h:336:0,
                 from /root/cphalcon/ext/phalcon.c:15:
/root/cphalcon/ext/phalcon/translate/adapter/csv.zep.h:10:26: warning: 'zephir_init_properties_Phalcon_Translate_Adapter_Csv' declared 'static' but never defined [-Wunused-function]
 static zend_object_value zephir_init_properties_Phalcon_Translate_Adapter_Csv(zend_class_entry *class_type TSRMLS_DC);
                          ^
make: *_\* [phalcon.lo] Error 1

**Comment by steffengy** [comment]
I do not know why people still try to compile phalcon under PHP7,
I said several times that it cannot work at this point.

**Comment by r3ndyipb** [comment]
hahaha.. i just curious with php7 performance combine with phalcon..

**Comment by masums** [comment]
@Rendy ariawindana

I have tested using PHP framework CodeIgniter with PHP 5.4 and 7.0 RC5. PHP
5.4 takes 650ms for single mysql query response and PHP 7.0RC5 takes 360ms

On Sat, Oct 31, 2015 at 11:54 PM, Rendy ariawindana <
notifications@github.com> wrote:

> hahaha.. i just curious with php7 performance combine with phalcon..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phalcon/zephir/issues/756#issuecomment-152755790>.
>



-- 
Md. Rashidul Hasan Masum <http://masums.com>
Software Architect
OnnoRokom Software Ltd. <http://onnorokomsoftware.com>
Mobile: 88018 11 41 95 57
